### PR TITLE
Add intersection

### PR DESCRIPTION
**Pull request by JakobBruenker**
First of all, thank you for this library, it seems to be exactly what I was hoping to find when I entered "typescript json decoding library" in Google.

I noticed that the library defines union types, but not intersection types, like `a & b`. I've found them quite useful for emulating inheritance.

By taking a look at [this stackoverflow answer](https://stackoverflow.com/a/59463385), I was able to come up with this:

```typescript
// intersectUnion<a | b> = a & b
type intersectUnion<U> = (U extends unknown ? ((_: U) => void) : never) extends ((_: infer I) => void) ? I : never;

// asObject<[a, b]> = { 0: {_: a}, 1: {_: b} }
type asObject<T extends unknown[]> = {[K in Exclude<keyof T, keyof []>]: {_: decodeType<T[K]>}};

// values<{0: a, 1: b}> = a | b
type values<T> = T[keyof T];

// fromObject {_: a} = a
type fromObject<T> = T extends {_: infer V} ? V : never;

// combine helpers to get an intersection of all the item types
type getProductOfDecoderArray<arr extends Decoder<unknown>[]> = fromObject<intersectUnion<values<asObject<arr>>>>;

export const intersection =
  <decoders extends Decoder<unknown>[]>(...decoders: decoders) =>
    (value: Pojo): getProductOfDecoderArray<decoders> => {
      const errors: any[] = [];
      const results: any[] = [];
      for (const decoder of decoders) {
        try {
          results.push(decode(decoder)(value));
        } catch (message) {
          errors.push(message);
        }
      }
      if (errors.length === 0) {
        // When not all cases are objects, the static type checking of the
        // intersection should ensure that we can safely return just the
        // result of the last case
        return results.some(res => typeof res !== 'object')
          ? results[results.length - 1]
          : results.reduce((acc, val) => ({...acc, ...val}), {});
      } else {
        errors.push(`Could not match all of the intersection cases`);
        throw errors.join('\n');
      }
    };
```

Note that if multiple cases create properties with identical keys, only the last one is kept. However, I don't think that's really a problem - after all, `union` has order dependency as well.

The PR includes this change as well as several test cases for it.

### CODE REVIEW

**Comment by tskj** [comment]
I'm glad you found the library useful! That's fantatstic!

Thank you for submitting this improvement. Don't know why I hadn't considered intersection types, but they seem like an obvious feature now that you point it out.

I want to fully understand what your solution does and how it works, I think I need a day or two to fully review it and I'll probably get back to you with questions before merging! Again thank you, this is great.

**Comment by tskj** [comment]
Alright, this is turning out to be even more complicated a feature than I first thought. I will have to look more at it this weekend when I have more time, However, a few points for consideration for now.

Did the npx tests pass on your end? I had to make a few changes to make them work, most notably change to `expectAssignable`, like the following:
```typescript
expectAssignable<{ a: 'foo' }>(intersection_decoder({a: 'foo'}))
```

Further, I am a bit skeptical to the test which checks for an object in the result. Is it really that simple? Also, only checking against `typeof res === 'object'` allows `null` (because they are considered objects in js).

I wouldn't mind the order dependency you mention, but I'm not really sure that is how intersections work in Typescript, and that's definitely not how the code you wrote behaves. If any of the decoders fail, it fails the entire `intersection`. Note I think this is the correct behaviour, but it doesn't quite match your idea of emulating inheritance. For example, if you had the following code

```typescript
intersection({ a: number }, { a: string })
```

you might expect that to be equivalent to `{ a: string }` if you consider the second `a` to "override" the first, in typical OOP terminology. Rather, this turns out to be `{ a: number & string }` which is further reduced to `{ a: never }`. Your implementation correctly models this by failing regardless of input. This is illustrated by the following test case (which fails)

```typescript
test('intersection', () => {
  const test_value = { a: 'test' };

  type intersection = decodeType<typeof intersect_decoder>;
  const intersect_decoder = intersection({ a: number }, { a: string });

  expect<intersection>(intersect_decoder(test_value)).toEqual(test_value);
});
```

I have pushed the corresponding code I used to test this with to [this branch](https://github.com/tskj/typescript-json-decoder/tree/feature/intersection). Not quite sure how to best collaborate on a PR from a fork like this?

Looking forward to hearing your inputs!

**Comment by JakobBruenker** [comment]
Thank you for your feedback!

> Did the npx tests pass on your end?

hm yes, I ran `npm test`, and it did (and does now) show all tests as passing (and doesn't if I add an incorrect test).
EDIT: Ah, this is true for the tests in e.g. `index.test.ts`, but not `index.test-d.ts`. Do I need to run something else to run those tests?

> I wouldn't mind the order dependency you mention, but I'm not really sure that is how intersections work in Typescript, and that's definitely not how the code you wrote behaves.

It could be that I expressed myself poorly, because the behavior you describe as what the code does and what's likely correct ("If any of the decoders fail, it fails the entire `intersection`", as well as `number & string == never`) *is* what I was aiming for.

The order dependency only comes into play when there are multiple compatible types, but they are objects that have the same field. E.g.
```typescript
console.log(intersection(
  {a: number},
  {a: field('a', n => number(n) + 1)}
)({a: 0})
```
will print `{a: 1}` rather than `{a: 0}`.

> but I'm not really sure that is how intersections work in Typescript

I don't know if that's really a meaningful comparison, since those only operate on types, whereas the order dependency is about differing values. Thus the type that `intersection` produces is not order-dependent, but the value is (just like with `union`).

> it doesn't quite match your idea of emulating inheritance

At the end of the day I suppose it's more important that this matches the bevavior of `&` rather than the behavior of inheritance - but that *is* how inheritance work in typescript:
```typescript
class A {
  a: number;
}

class B extends A {
  a: string;
}

/*
Property 'a' in type 'B' is not assignable to the same property in base type 'A'.
  Type 'string' is not assignable to type 'number'.
*/
```

I'll leave it here for now, I'll address the 'object' check in a separate comment.

**Comment by tskj** [comment]
Not sure what is happening with the tests then, maybe there is a misconfiguration in `package.json`. When you run `npm run test` it fist invokes `jest` which runs the `index.test.ts` tests, and then it invokes `tsd`, which is a utility that tests only types (which are the `index.test-d.ts` tests). Not quite sure how that works, but apparently it only parses the Typescript code and validates the AST at a type level. Maybe `tsd` isn't being installed properly by `npm i`?

Anyway, then it seems I was misunderstanding you! I think we agree on all points.

The type `{ a: number } & { a: string }` is essentially impossible to instantiate in regular Typescript (you might think it gets reduced to `{}`, but that doesn't seem to be the case). I think this implementation correctly models this behavior.

However, I am concerned by another case, which is more tricky. Namely the type `{ a: number } & { b: boolean }`, which really is equivalent with `{ a: number; b: boolean }`. I'm not quite sure how this would even be possible to implement with decoders, because both decoders would separately fail (right?), but should succeed together.

If you only use `intersection(a, b, c)` in a way where `c` is a subtype of `b` and `b` is a subtype of `a`, that is, you have an inheritance hierarchy in increasing order of specificity, this would work out. But the intersection operator `&` in typescript is more general than that, and although I think demanding an order dependency might be fine, there are cases such as the one I mentioned above, where there is no order that works, because neither `a` nor `b` are subtypes of each other; but they should still form a valid type.

Maybe this is why multiple inheritance is eschewed in other languages, hah? Tricky problem for sure, but I'm sure we'll find a solution!

**Comment by JakobBruenker** [comment]
I'll see if I can figure out what's happening with the tests later.

> because both decoders would separately fail (right?)

```typescript
console.log(intersection({a: number}, {b: boolean})({a: 123, b: false}));
```
prints `{a: 123, b: false}`.  
Which I suppose comes down to how you implemented record decoders, where they don't mind having additional fields in the input data.

One thing I found out though is that at the moment,
```typescript
type Weird = number & {a: number};
const weirdDecoder = intersection(field('a', n => number(n) + 1), {a: number});
const w: Weird = weirdDecoder({a: 0})
console.log(w);
```
prints `{a: 0}`, where I think it should be failing, since I don't think you can have a value of type `number & {a: number}`.

So you were right, it's not quite as simple as the `'object'` check - I'll have to think about this.

> I have pushed the corresponding code I used to test this with to [this branch](https://github.com/tskj/typescript-json-decoder/tree/feature/intersection). Not quite sure how to best collaborate on a PR from a fork like this?

Should I make another PR against that branch if I have additional changes for this feature?

**Comment by JakobBruenker** [comment]
> Should I make another PR against that branch if I have additional changes for this feature?

Or instead I could just merge your changes from that branch into this PR

**Comment by tskj** [comment]
Ah I see I wasn't careful enough in understanding the implementation, my apologies for that; combining the objects by spreading does actually address my concern.

Maybe it's useful to split the discussion into the two cases you had already identified in code: only objects, and a mix. For the only objects case I am now convinced it does actually behave more or less identically to `&` for the cases that matter.

As for the case where there is a mix, most of them are probably invalid, such as the Weird example you give. I agree primitives and objects make no sense together. The same probably goes for all other primitives mixed with each other, the only case I can see is primitive literals like `false` and `"foo"`, which when intersected with their parent types like `boolean` and `string` (respectively) should narrow to the literal. So maybe it's fair to claim that there really are only two cases, either they narrow to the literal, or they narrow to `never`? Not sure if that observation helps us though.

Also another thing to consider is arrays, but that might be solved directly because they are treated as objects with keys `'0'`, `'1'`, etc.? Maybe also tuples, but that might be ill defined in any case.

Yes feel free to merge my changes into your branch!

**Comment by JakobBruenker** [comment]
Arrays and tuples is a good point; we can't just treat them as objects, since `Array.isArray([1,2,3]) !== Array.isArray({0: 1, 1: 2, 2: 3})`.  
Maybe instead of combining the objects by spreading, we should be using `Object.assign`.
Though then the question would be which object to choose as the base object - I suppose the natural choice would be the first one.

Also, there's this example
```typescript
  const tuple1: [number, number] = [1,2];
  const tuple2: [number, number, number] = [1,2,3];
  const decoder: DecoderFunction<never> = intersection(_ => tuple1, _ => tuple2);
```

which compiles and runs - but shouldn't. I don't have a good idea on how to solve this at the moment, but if there is a feasible solution, it probably involves the type of `intersection` rather than anything at runtime, since we can't distinguish between arrays and tuples at runtime, as far as I'm aware.

**Comment by tskj** [comment]
Actually I think `{ ...acc, ...val }` compiles to something like `Object.assign({}, acc, val)`, so I don't really mind the spread operator - I think the spread operator is the clearest solution, arrays probably need special handling in either case.

We might have uncovered a much broader issue here, which is that decoders can create arbitrary values. We need to handle any custom decoder, which might be decoding Dates, Maps, Sets or whatever really. I think I need to check out how this is handled by Typescript natively before I can have an opinion on how that should work.

Either that or we could detect anything other than the clear cut cases somehow at the type level and fail them there.

**Comment by JakobBruenker** [comment]
Yeah, a related question is whether it's okay to discard any information about prototypes, e.g., say you have
```typescript
class B {
  test() {
    console.log('B');
  }
}
```
and a `intersection` that produces `new B()` in one branch, is it okay if the resulting value does not have `test()`?

It makes things a bit more complicated, but it does look like Typescript's `&` takes them into account. So I suppose `intersection` should, as well, otherwise it isn't typesafe. The solution here would be to (before writing `{...acc, ...val}`) doing something like
```typescript
{...acc, ...Object.getOwnPropertyNames(Object.getPrototypeOf(val)).map(prop => {[prop]: val[prop]})}
```
(Edit: this isn't quite right, but it's close)
(and the same with `Object.getOwnPropertySymbols`).
I was hoping that would automatically take care of the Array issue, but alas, Arrays are special, so I suppose need a special case regardless of whether or not this ends up being part of it.

But I believe this would take care of Map, Set, Date, etc.

**Comment by JakobBruenker** [comment]
One way to handle the Array special case - ~~which would require using `Object.assign`, which, I agree, is less clear than the spread operator~~ * - would the to check whether `Array.isArray` is true for any of the results, and start with `[]` if so, and `{}` otherwise.

Then, adding the properties of the other results via `Object.assign` will preserve the `isArray`ness.

As far as I can tell there are no other types with a method similar to `Array.isArray`, so no other special object cases (apart from `null`, not entirely sure how to handle that yet).

*Edit: Actually, maybe we can keep the spread operator and just have one final `result = []; Object.assign(result, acc)` at the end if there were any arrays.

**Comment by tskj** [comment]
I am inclined to say that any code that wants to do complicated `class` things and further intersect them, at the decoding stage, is maybe not worth supporting. For example `Map` and a primitive makes no sense, possibly something like `Map & Set` miight make sense, although it is a stretch, but certainly "adding" a property to a `Map` is something you might want to do.

The question then is, can we detect this at the type level and fail it (at compile time)? That way we could release this limited functionality (which arguably is the important functionality), and optionally expand our capability to support arbitrary intersections in the future without breaking backwards compatibility. I have no issue with supporting a limited subset of the `&` operator, or even our own (possibly different) interpretation of it - as long as what we implement ends up being type safe and soundly typed (and intuitive to use for common cases).

That being said, if we manage to come up with a solution, like the one you proposed with `ownProperty`, that is certainly cool! With a sufficient test coverage, it would be good to cover more usecases as long as we don't complicate the basic (and more important) usecases.

As far as arrays go, I don't think the spread operator with `[]` is correct, as that essentially operates as `concat`.

First consider decoding an array twice `number[] & numbe[] == number[]`, which in decoders would be `intersection(array(number), array(number))([ 1, 2, 3 ]) == [ 1, 2, 3, 1, 2, 3 ]`. The array elements would end up duplicated, which is accidentally typesafe, but (I think) clearly incorrect behavior.

Second consider the case of two tuples `[string, number] & [number, number] == [never, number]`, this would also duplicate the elements if it worked, but more to the point it would fail on one of the elements, while a correct implementation would produce a `never` for the first element, and a `number` for the second element.

*(An interesting caveat to the previous statement is that while it seems impossible to produce a `never`, since it is impossible for the client code to use or inspect such a value, we can actually do this, e.g. by providing `null as never`. This actually gives the client code the possibility to access the second element, and therefore the possibility for a wider range of correct programs to exist. But this might be of more academic interest than anything practical, I don't know.)*

This seems incredibly challenging to detect reliably, so the pragmatic thing to do might be to not support these cases at all. This is turning out to be a ridiculous rabbit hole, but I'm learning quite a lot about how Javascript works, which is good, and I am confident we will figure this out!

**Comment by JakobBruenker** [comment]
> For example Map and a primitive makes no sense, possibly something like Map & Set miight make sense, although it is a stretch, but certainly "adding" a property to a Map is something you might want to do.

One thing that might come up (although I don't necessarily see myself doing it) is to have something like custom classes `A` and `B`, and then write a decoder of type `A & B`, thus effectively getting multiple inheritance, with the result having all the methods of `A` and of `B`.

More practically though, I kind of think that supporting this is easier than figuring out how to prevent it at the type level (assuming that's possible at all). Though that may just be because I have a solution in mind for the former, but not for the latter. Will still need to do *something* at the type-level anyway, to prevent things like `intersection(_ => tuple1, _ => tuple2`).

> as long as what we implement ends up being type safe and soundly typed (and intuitive to use for common cases)

these seem like sane priorities

> As far as arrays go, I don't think the spread operator with [] is correct, as that essentially operates as concat.

Hmm, I'm not sure what exact combination of array and spread operator we're talking about. To be clear,  
(A) `({...[1,2,3], ...[1,2,3]}) = { 0: 1, 1: 2, 2: 3 }` &ndash; no duplication, `isArray` is false  
(B) `({...[], ...[1,2,3]}) = { 0: 1, 1: 2, 2: 3 }` &ndash; same as (A)
(C) `Object.assign({}, [1,2,3]) = { 0: 1, 1: 2, 2: 3 }` &ndash; same as (A)
(D) `Object.assign([], [1,2,3]) = [1,2,3]` &ndash; `isArray` is true
(E) `Object.assign([], { 0: 1, 1: 2, 2: 3 }) = [1,2,3]` &ndash; same as (D)
Edit:
(F) `[...[1,2,3], ...[1,2,3]] = [1,2,3,1,2,3]` &ndash; this does duplicate, but I think we only need `...` inside `{}` anyway.

> This is turning out to be a ridiculous rabbit hole

Indeed, it's surprising how complex this is considering it's just the dual to `union`. I'm also learning quite a bit about JS here.

**Comment by tskj** [comment]
So it was definitely F I was thinking of! Unfortunately D and E are equivalent to F as far as I know, it only happens to work because the starting array is empty (a necessity because `Object.assign` mutates it, but as soon as you want to combine two arrays with data, you get the same problem). So I don't know how to solve this one, without losing the `isArray` property.

An alternative is to fail the decoder in some way. I agree it seems difficult to avoid at the type level, and I don't have a solution in mind either, so we might have to give up on that. However, we could still avoid being type unsafe by detecting the array / tuple case (using `isArray`) at runtime and failing the decoder with a custom error message. This would give the appearance to the user of working from the type signature, but I don't think we would be violating type safety by not supporting it (again, as long as we fail hard by throwing, instead of returning something that isn't an array when the signature says we will be returning an array). This also gives us room to actually implement some kind of solution in the future without breaking.

The `hasProperty` solution is very nice, I encourage you to see what you can do with it!

**Comment by tskj** [comment]
Oh and about the `'object'` test, I think a solution that covers everything we have discussed is to test against `'object'`, but also handle `null` and things that are `isArray` (which will both return true for `'object'`) specially, and maybe as I suggested in my previous comment, just reject them all of them outright. This also means we would not have to special case and return (e.g.) the last match, but could in fact combine everything just with a normal spread.

I'm open for other solutions as well, but I think this in sum addresses all my concerns at least.

**Comment by JakobBruenker** [comment]
> Unfortunately D and E are equivalent to F as far as I know

Looks like it works:
`Object.assign([1,2,3], [1,2,3]) = [1,2,3]`

However, I wasn't planning to do that anyway - my plan was to use `{...acc, ...val}` as in the current commit, and then at the end run
```typescript
return Object.assign(doWeNeedAnArray ? [] : {}, result}
```
which should work even if it only works with `[]`.

> just reject them all of them outright

i.e. reject it outright if anything isn't an `'object'`? That could work, but it would mean we give up on somewhat reasonable usecases like `(number | string) & (number | boolean)`.

Or reject if there's a mix of `'object'`s and non-`'object'`s? (In that case we'd still need a special case of returning the last match to avoid converting a value of something like `(number | string) & (number | boolean)` to an object via spreading operator).

In any case, I will try to see if I can find out what's going on with the tests and afterward see how far I can get with `ownProperty` later today.

**Comment by tskj** [comment]
Okay I thought I understood all of the `Object.assign` cases, but apparently not. Would doing the final assign like you had intended preserve other properties on the object if it "goes into" an array?

I did mean reject everything else yes, but you do make a good point. I wonder if there is a somewhat reasonable example where the decoders don't produce the exact same value but still make sense. Because in your example there would be two `number` decoders succeeding and producing the same value. Clearly if every such case produces the same value it doesn't matter which one we pick, but I'm curious if there is a more degenerate case that type checks.

I think a mix of objects and primitives is a case we can safely reject at least?

This also reminds me that we need to handle the case of someone calling `intersection()` without parameters! Not sure how e.g. `union()` handles this, I would check if I wasn't on my phone right now, but maybe it just infers the type to `never`?

**Comment by JakobBruenker** [comment]
> Would doing the final assign like you had intended preserve other properties on the object if it "goes into" an array?

Yep, so, for example `Object.assign([1,2,3], {a: 'foo'}).a = "foo"`. Just ends up being an array with additional properties.

> But I'm curious if there is a more degenerate case that type checks.

One non-changing example of `number & number` (without unions) that I can think of is combining validators:

```typescript
nonnegative: Decoder<number> = n => {
  const num = number(n)
  if (n < 0) throw 'too small'
  else return n
}
even: Decoder<number> = n => {
  const num = number(n)
  if (n % 2 !== 0) throw 'too odd'
  else return n
}
nonnegative_even = intersection(nonnegative, even)
```

(edit: I suppose you don't need `intersection` for that though, you can just do `n => {nonnegative(n); return even(n)}`). But I think if you have a validators that produce unions for some reason, you could actually benefit from intersection here, since it'd have a narrower inferred type for the result.)

As you say, this gets weird if the value isn't always the same - but you might reasonably want something like
```typescript
toAbsolute: Decoder<number> = n => Math.abs(number(n))
absolute_even = intersection(even, toAbsolute)
```

Though doing that in practice with the order-dependency would be confusing I think, so it's not exactly something one should be encouraging I suppose.

One way to sidestep this entirely would be to require that each branch returns the same value, or the same value for all fields that actually exist, but then for nested objects that would be a small performance-hit, since you have to traverse them again - and not all objects might reasonably be comparable (e.g. functions)

> I think a mix of objects and primitives is a case we can safely reject at least?

I believe so, same if the branches produce a mix of different primitive types.

> without parameters

`union()`'s type is indeed `never`. Seems sensible, since `never` is the identity element of `|`, so `a | never == a`.

The obvious choice as unit element for `&` is `{}`, and that's what my commit produces at the moment for an empty `intersection`. I actually thought this wouldn't work for primitive types, but it turns out, `a & {} == a` even if `a` is a type like `number`. (there are exceptions though, `{} & null` and `{} & undefined` is never. I suppose that's probably tolerable though.)

**Comment by JakobBruenker** [comment]
I figured out why the tests didn't work - I had to run `npm run build` first, otherwise I got ``The type definition `dist/index.d.ts` does not exist. Create one and try again.`` below the `jest` results.

**Comment by tskj** [comment]
Oh that's weird about the tests, probably badly configured and I hadn't noticed because I've run npm build before. Although kind of strange, because I don't think it reruns old tests even if you don't build after changing a type test? I'll have to look into it.

I think the non negative and even case is very cool, og course it would be much cooler still if the type system remembered that information and didn't reduce it to just `number`, but that's neither here nor there. The case with `toAbsolute` feels very different because you're actively transforming one of them, and seems like it should rather be `n => toAbsolute(even(n))`.

Sounds like maybe a reasonable compromise is this:

- a mix of primitives and objects is out
- all primitives means they should all be equal (equality test is simple)
- anything else is considered an object, and they are combined using some kind of spread scheme

If it's not an object either, like a function - well I don't know, we probably shouldn't allow that, but I don't know if we can catch all those cases to ban them either?

**Comment by JakobBruenker** [comment]
I've added the implementation of adding the methods and special casing array. I haven't really tested it yet, so it might still be wrong. The basic approach (slightly different from what I wrote above) is:
 - get prototype of each result
 - combine prototypes
 - set prototype of the result to this combined prototype

I've also implemented the compromise you mentioned about testing only primitives for equality. I think treating objects and primitives differently is a disadvantage, but it does seem like an okay compromise (what with all solutions having some disadvantage).

What's still missing now is:
- null/undefined treatment (I haven't yet thought very much about what the right way to handle these is)
- exclude certain primitives at the type level (symbol and function?)
- Don't allow the `intersection(_ => tuple1, _ => tuple2)` example on the type level, somehow
- add tests for all of these changes

**Comment by tskj** [comment]
Very good! I will help adding some test coverage when I have time, probably the weekend.

I think there is some sense in treating objects and primitives differently, because something like `intersection({ a: string }, { b: number })` actually creates a new type that is a subtype of both, while doing something similar for primitives is really only possible if the resulting subtype _is_ one of the mentioned types. At least this is my intuition, and I can't think of a case where primitives behave differently.

Null and undefined I believe can just be treated the same as other primities maybe?

No idea how to handle symbols or functions; functions especially will be treated as decoders by the type system, so it's not really something you can reasonably decode to anyway.

I think maybe tuples will end up being coalesced with the array handling, and I wouldn't be surprised if the types end up being inferred as just an array (of which a tuple is a subtype in typescript?). If that is what happens, I think it's an ok outcome. It would be cool to support tuples for sure, but it also seems like an alright compromise to treat them as arrays. This also keeps the door open for a better implementation later without breaking, I would think.

Again, very nice work! Look forward to getting this merged.

**Comment by tskj** [comment]
I have added some test cases and pushed to my branch. Most tests run fine, however there are two tests you'll see I had to comment out, which I think demonstrate that arrays and tuples actually aren't viable. Thinking more about how that is supposed to work, it becomes more and more clear that it sadly looks like a pretty insurmountable problem. Luckily I think it will be easy to detect at runtime and fail like we have discussed, and I suppose fairly easy to explain in the documentation that intersection only works for pure objects or pure primitives (and unions thereof).

I'm starting to think that the core design of the library, where a `Decoder<T>` is a function from `Pojo` to `T`, maybe was a mistake. If decoders were represented as a data structure, which was passed to a decoding function, like `decode(decoder, pojo)`, we would now have the opportunity to traverse that structure and re-arrange it to an equivalent decoder which could handle this case. It would still be difficult of course, but I think as it stands it's actually impossible.

I'm curious what your thoughts are on the issue are!

**Comment by JakobBruenker** [comment]
Thanks, I've added the commits to the PR, I'll let you know what I think once I've looked at it

**Comment by JakobBruenker** [comment]
So if I'm seeing this right the first test ("intersection of arrays of objects") fails to compile on
```typescript
  expect(intersect_decoder(test_value).map(x => x.b)).toEqual([ 1, 3 ]);
```

with `Property 'b' does not exist on type '{ a: string; }'`

the same thing happens without decoder when writing

```typescript
type A = {a: string}[] & {b: number}[]

const foo: A = [{a: 'test', b: 45}, {a: 'hello', b: 42}]

foo.map(x => x.b)
```

is there a reasonable explanation for that error? It kind of looks like a bug in typescript to me, but maybe I'm missing something.


With the second test, I think that actually demonstrates a more general problem that comes from overwriting object properties, and isn't a problem unique to arrays/tuples - the same thing happens when using `{ 0: {a: string; b: number}, 1: number }` instead of `[ {a: string; b: number}, number]`.

I think the solution here is that we have to traverse the objects after all, and at each level, combine the properties like is being done only at the top level at the moment. (And if we're traversing anyway, I suppose we might as well check that identical fields contain equal values, like we do with primitives.)

You're suggesting that we would have to traverse the decoder, but I think it's enough to traverse the results the decoders produce.

**Comment by tskj** [comment]
Wow yeah that really does look like a typescript bug. 

Huh, traversal it is then. I think you're right that traversing the objects might be enough actually, thinking about it, it seems also much easier than trying to reduce a decoder data structure to something simpler (which might not even be possible in all cases).

I think more test cases are in order.

**Comment by JakobBruenker** [comment]
Re: the typescript bug, I found this issue https://github.com/microsoft/TypeScript/issues/41874  
TL;DR: It's a design limitation and will not be fixed.

(As far as this patch goes, I don't think anything really needs to be changed because of it?)

**Comment by tskj** [comment]
Ah, good find! Coming from a FP mindset, I tend to think of arrays as immutable, and therefor covariant. But of course, being mutable they aren't really covariant or contravariant.

So `{ a: string; b: number }[]` isn't really a subtype of `{ a: string }[]` because then you could add `{ a: string }` objects to the first array, which is clearly unsound.

Either way, I will fix the test case then. Maybe iterating over arrays and combining objects at matching indices will be sufficient then?

**Comment by JakobBruenker** [comment]
I think so, I'll implement that later today and we'll see if we find any more cases that break it

Good observation regarding the variance, I also tend to assume immutability by default

**Comment by JakobBruenker** [comment]
I've implemented the recursive combining of objects and added a couple of tests.

The "intersection of compatible tuples" test you added now passes.

The prototype thing was actually slightly more complicated than what I committed before because of inheritance - the prototypes of the prototypes (and so on) need to be combined as well. (See test "class intersection".)

**Comment by JakobBruenker** [comment]
I think we need special cases for Map and Set as well - unless they're not worth supporting, I can't imagine it'd be a very common use-case. The problem there is that (unlike arrays) they don't store their elements as object properties.

**Comment by tskj** [comment]
Ah I see, yeah anything that has private data accessible only through methods will be impossible I guess.

I wonder if we should detect such cases and just fail the decoder? That kind of solves everything we can't to while explaining to the user that it won't work, rather than silently giving wrong answers.

**Comment by JakobBruenker** [comment]
To be honest I didn't know Javascript (as opposed to TS) had private fields until just now. Detecting whether something is a Map/Set is easy, detecting whether something has private fields is quite a bit harder.

[The FAQ](https://github.com/tc39/proposal-class-fields/blob/main/PRIVATE_SYNTAX_FAQ.md) says
> private fields are purely internal: no JS code outside of a class can detect or affect the existence, name, or value of any private field of instances of said class without directly inspecting the class's source

Looking at the source code *is* possible, by looking at `prototype.constructor.toString()`, but it doesn't seem easy to make finding private fields in there robust :-/

I suppose an alternative would be to have the decoder fail on any custom class, and instead only allow objects if they have the same prototype as `{}` (as well as primitives).
(On the plus side, that would simplify the code by getting rid of the current prototype handling)

**Comment by tskj** [comment]
I didn't really know that either, but also any method can capture something in a closure, effectively making it super private.

So yeah, I think maybe disallowing custom classes is the way to go, decoding objects (`{}`) and arrays (`[]`), as well as primitives, really is the central part of Json decoding anyway. However, I think we still maybe need to be on the lookout for functions and disallow them inside objects as well due to the above reasons.

**Comment by JakobBruenker** [comment]
Okay, I've pushed a version that removes support for custom classes. Still have to do some finishing touches tomorrow, if we don't discover more issues that need to be resolved.

**Comment by tskj** [comment]
Brilliant! I think maybe finally we got it.

**Comment by JakobBruenker** [comment]
Okay, I've added a few lines to treat null essentially as a primitive type - the same was already true for `undefined`, since it actually is one. I also added a couple more tests.

One thing we talked about last week that I didn't do is forbid intersection like e.g. `[number, number] & [number, number, number]` on the type level - but the difference to last week is that the decoder now fails at runtime if something like this is attempted to be decoded, since the values won't be equal. Do you suppose that's good enough?

(More generally this also means that intersections of arrays must produce the same array - which could be seen as either a good thing or a bad thing I suppose. I didn't program that in explicitly, it just arose naturally since their `length` properties aren't equal.)

**Comment by JakobBruenker** [comment]
By the way, I noticed that ~~strictNullChecks doesn't seem to be on in the test suite~~ - which doesn't make a difference for running the tests of course, just stumped me for a bit when typescript claimed that the result of `(number | null) & (string | null)` is `never` as opposed to `null`. Just pointing this out in case you want to enable it.

Edit: Nevermind! It is actually enabled. VS Code just doesn't have it enabled in the tests for the typescript settings it uses for tooltips etc. Maybe it can't figure out what the correct tsconfig is. Anyway, seems like this is VS Code problem rather than a testsuite problem.

**Comment by tskj** [comment]
I assume you mean `(number | null) & (string | null)` was supposed to be `null` and not `string`? Anyway, annoying that VS Code doesn't follow the settings, don't know what to do about that.

So it wouldn't be unreasonable I think to consider a shorter array to be a subtype of a longer array, and by extension `[number, number] & [number, number, number]` to be equivalent to `[number, number]`, but I don't know if that's even possible to implement, so just failing seems completely fair. I'll take a close look after work!

**Comment by JakobBruenker** [comment]
>  was supposed to be `null` and not `string`

yep, that was a typo

> `[number, number] & [number, number, number]` to be equivalent to `[number, number]`

If you mean the tuple types (rather than arrays with length 2 and 3), this doesn't work, since typescript's type system reduces `[number, number] & [number, number, number]` to `never`.

It could be desirable to combine arrays of different lengths (with the result being the longest) though, but that seems hard to reconcile with intersections of tuples (since they're the same at runtime).

**Comment by tskj** [comment]
Ah I didn't know tuples did that. In that case I think the current implementation is the best solution! 

Also you're right, I definitely got that backwards, of course the longest array has to be a subtype of the shorter. This is definitely too tricky and esoteric to deal with.

**Comment by tskj** [comment]
I think I had the same issue with the nullable/optional decoders in VS Code, so I added a (type) test case for it. Also fixed my out-commented test case, and fixed a bug in another.

I'm very happy with how this has turned out! Give me a thumbs up when you're happy on your end and I'll merge it.

**Comment by JakobBruenker** [comment]
Good to hear! I merged your changes into the PR. No more changes needed from my side.

**Comment by tskj** [comment]
Alright then it's released as `1.0.7` on NPM. Thanks so much man! Really great work!

**Comment by JakobBruenker** [comment]
Nice! Thanks for all your comments on this PR.

**Comment by tskj** [comment]
Hey would you mind shooting me an email at epost@tarjei.org? I couldn't find any contact info on your github page. I promise I won't pester you or anything!

**Comment by JakobBruenker** [comment]
I forwarded this github notification thread to your email from my address
